### ISSUE TITLE
Reusing and initalization of Ops/Computers

### ISSUE THREAD

**Issue by dietzc** [imagej/imagej-ops#171]
Hi @ctrueden and @hinerm,

I can't remember if we discussed this already, but in some situations it is _very_ useful if 

* there would be a `initialize` call on an `Op` right after the `Op` has been created, especially if one wants to re-use the same operation over and over again (for example if the `Op` actually is a `Computer`, one may want to instantiate the `Computer` once and re-use it over and over again).

* one could ask the `OpService` to return only `Op`s which are `Function`s or `Computer`s.

I know that `Module`s have a life-cycle covering all that stuff, but it seems that `Module`s do way more than just `initialization`. 

What do you think?

**Comment by dietzc** [comment]
actually these are two issues in one. the second issue (`ops.computer(...)` or `ops.function(...)`) is really need in situation where the (often auto-generated) interface is not sufficient. For example see: https://github.com/imagej/imagej-ops/blob/7fde3628f549d158ccd7d1e753180a487a6c85f1/src/main/java/net/imagej/ops/copy/CopyIterableInterval.java#L75. 

This is also related to: https://github.com/imagej/imagej-ops/issues/164

**Comment by ctrueden** [comment]
Regarding initialization: yes, `Module` already has a mechanism for it, which we could easily make to work with `Op`s also.

> one could ask the `OpService` to return only `Op`s which are `Function`s or `Computer`s.

As discussed on Gitter, my reservation with this is that the `OpService` already has a general mechanism for restricting the type of a desired op: `<OP extends Op> OP op(Class<OP> type, Object... args)`.

This gives you exactly the needed type of op, and no other type (or null if no match, of course).

There are issues with generics for `Function` and `Computer` though—there is no obvious way to restrict the `I` or `O` of e.g. the returned `Function<I, O>`. I'm still trying to understand if you see other problems _besides_ that one, or if it would be enough to have:

```java
<I, O, OP extends FunctionOp<I, O>> OP function(Class<OP> opType, Class<I> inputType, Class<O> outputType, Object... args);
<I, O, OP extends ComputerOp<I, O>> OP function(Class<OP> opType, Class<I> inputType, Class<O> outputType, Object... args);
```

Furthermore, we could add a new `InputOutputOp` interface which unions `InputOp` and `OutputOp`, and which is a supertype of `ComputerOp` and `FunctionOp`, and then we'd only need one new sig:
```
<I, O, OP extends InputOutputOp<I, O>> OP op(Class<OP> opType, Class<I> inputType, Class<O> outputType, Object... args);
```

I quickly tried to hack this into the `OpService` like that, but it is rather involved. So before I proceed further: is this what you are looking for?

**Comment by dietzc** [comment]
Nearly, I think I need more something like

```
<I, O, OP extends Function<I, O>> OP op(Class<? extends Op> opType, Class<O> outputType,
     Class<I> inputType, Object... args);
```

which fails if there is no `Op` which is a `Function<I,O>`. Does this make sense?
The main idea is, that I want to make sure, that I get something back which is a `Function` or `Computer` for further processing (and usage of e.g. O compute(I input)). 

The suggestions you made are close, but imply that I have a general interface for my `OP` which is a `Function`.

**Comment by ctrueden** [comment]
Unfortunately, the signature you propose cannot work because `OP` is never passed anywhere as part of the input parameters.

However, we could maybe do something like:
```java
public <I, O, OP extends InputOutputOp<I, O>> OP op(Class<? extends Op> opType,
    Class<OP> returnType, Class<I> inputType, Class<O> outputType, Object... args)
{
   ...
}
public <I, O> FunctionOp<I, O> functionOp(Class<? extends Op> opType,
    Class<I> inputType, Class<O> outputType, Object... args)
{
    return op(opType, FunctionOp.class, inputType, outputType, args);
}
<I, O> ComputerOp<I, O> computerOp(Class<? extends Op> opType,
    Class<I> inputType, Class<O> outputType, Object... args)
{
    return op(opType, ComputerOp.class, inputType, outputType, args);
}
<I, O> HybridOp<I, O> hybridOp(Class<? extends Op> opType,
    Class<I> inputType, Class<O> outputType, Object... args)
{
    return op(opType, HybridOp.class, inputType, outputType, args);
}
<I, O> InplaceOp<I, O> inplaceOp(Class<? extends Op> opType,
    Class<I> inputType, Class<O> outputType, Object... args)
{
    return op(opType, InplaceOp.class, inputType, outputType, args);
}
```
Although there are almost certainly issues with nested generics there which would require some (ugly but safe) casts.

I guess this is what you were originally asking for: a dedicated method for each core interface...

**Comment by dietzc** [comment]
Yes, you are right. Won't work. However, we could do something like:
```java
<I, O, OP extends Op> OP op(Class<? extends OP> coreInterfaceType, Class<? extends Op> opType, Class<O> outputType, Class<I> inputType, Object... args);
```

Anyway. This all boils down to what you suggested. Do you think its a good idea? From my point of view its really useful. Maybe we can also put these methods in their own namespace `access() or so`, such that the `OpService` API  doesn't have too many methods.

**Comment by ctrueden** [comment]
OK, I'll try to finish the `input-output-ops` branch during the hackathon. Hopefully it will be a very short project. :pray:

**Comment by ctrueden** [comment]
Progress on the [op-chains](https://github.com/imagej/imagej-ops/compare/op-chains) branch. Still needs testing.

**Comment by ctrueden** [comment]
Thoroughly addressed with #202! Thank you hackathon!

**Comment by dietzc** [comment]
:+1:
